load-test-framework/dotnet_src/Google.PubSub.Flic/Program.cs

```diff
@@ -0,0 +1,174 @@
+                $"CPU scaling:{startRequest.CpuScaling}");
+            var now = DateTime.UtcNow;
+            stats.Init(now, startRequest.IncludeIds);
+            var waitDuration = startRequest.StartTime.ToDateTime() - now;
+            if (waitDuration > TimeSpan.Zero)
+            {
+                await Task.Delay(waitDuration);
+            }
+            switch (startRequest.ClientOptionsCase)
+            {
+                case StartRequest.ClientOptionsOneofCase.PublisherOptions:
+                    await Publish(startRequest, stats);
+                    break;
+                case StartRequest.ClientOptionsOneofCase.SubscriberOptions:
+                default: // no option is sent for subscribe
+                    await Subscribe(startRequest, stats);
+                    break;
+            }
+        }
+
+        private static async Task Publish(StartRequest startRequest, Stats stats)
+        {
+            var topicName = new TopicName(startRequest.Project, startRequest.Topic);
+            var clientCreationSettings = new PublisherClient.ClientCreationSettings(
+                clientCount: (int)Math.Max(Environment.ProcessorCount * startRequest.CpuScaling, 1));
+            var settings = new PublisherClient.Settings
+            {
+                BatchingSettings = new BatchingSettings(
+                    elementCountThreshold: startRequest.PublisherOptions.BatchSize,
+                    byteCountThreshold: null,
+                    delayThreshold: startRequest.PublisherOptions.BatchDuration.ToTimeSpan())
+            };
+            var client = await PublisherClient.CreateAsync(topicName, clientCreationSettings,  settings);
+            int msgsSent = 0;
+            double rate = startRequest.PublisherOptions.Rate;
+            var startTime = DateTime.UtcNow;
+            var endTime = startTime + startRequest.TestDuration.ToTimeSpan();
+            var dataBytes = new byte[startRequest.PublisherOptions.MessageSize];
+            var rnd = new Random(1);
+            rnd.NextBytes(dataBytes);
+            var data = ByteString.CopyFrom(dataBytes);
+            var msgsInFlight = 0;
+            const int MaxQueued = 1_000_000;
+            while (true)
+            {
+                while (true)
+                {
+                    // Inner loop to throttle publish rate on both requested publish rate, and local queue.
+                    var now = DateTime.UtcNow;
+                    if (now >= endTime)
+                    {
+                        stats.End(now);
+                        return;
+                    }
+                    var expectedMsgsSent = (now - startTime).TotalSeconds * rate;
+                    if ((rate == 0.0 || msgsSent < expectedMsgsSent) && Interlocked.Add(ref msgsInFlight, 0) < MaxQueued)
+                    {
+                        break;
+                    }
+                    await Task.Delay(TimeSpan.FromMilliseconds(1));
```

Comment:
Please don't sleep loop here, you have access to the information you need to determine when you have to wait.

For the rate case, you can calculate how long to wait until the next token is available.

For the outstanding case, you can use something like https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=netframework-4.8 with WaitAsync to limit the number outstanding.